packages/protocol/contracts/signal/SignalService.sol

```diff
@@ -137,18 +145,30 @@ contract SignalService is AuthorizableContract, ISignalService {
             Hop memory hop = p.hops[i];
             if (hop.stateRoot == stateRoot) revert SS_INVALID_HOP_PROOF();
 
-            bytes32 label = authorizedAddresses[hop.relayerContract];
+            bytes32 label = authorizedAddresses[hop.relay];
             if (label == 0) revert SS_HOP_RELAYER_UNAUTHORIZED();
 
             uint64 hopChainId = uint256(label).toUint64();
 
-            verifyMerkleProof(
-                stateRoot, hopChainId, hop.relayerContract, hop.stateRoot, hop.merkleProof
+            verifiedStorageRoots[hopChainId][hop.relay][hop.stateRoot] = verifyMerkleProof(
+                stateRoot,
+                hopChainId,
+                hop.relay,
+                hop.stateRoot,
+                hop.merkleProof,
+                verifiedStorageRoots[hopChainId][hop.relay][hop.stateRoot]
             );
             stateRoot = hop.stateRoot;
         }
 
-        verifyMerkleProof(stateRoot, srcChainId, app, signal, p.merkleProof);
+        verifiedStorageRoots[srcChainId][app][stateRoot] = verifyMerkleProof(
```

Comment:
In my opinion, the underlying LibTrieProof library shall know nothing about the whole caching idea.

Reply:
I'd improve the ICrossChainSync interface and its underlying implementation to support caching of state roots from multiple chains.